components/org.wso2.carbon.identity.conditional.auth.functions.http/src/main/java/org/wso2/carbon/identity/conditional/auth/functions/http/AbstractHTTPFunction.java

```diff
@@ -73,65 +82,248 @@ public AbstractHTTPFunction() {
+                    LOG.error("Request URL does not match with the allowed domain list. Request Url: " +
+                            endpointURL);
+                    asyncReturn.accept(context, Collections.emptyMap(), Constants.OUTCOME_FAIL);
+                } else {
+                    Pair<RetryDecision, Pair<String, JSONObject>> result = executeRequest(request, endpointURL);
+                    if (result.getLeft().shouldRetry()) {
+                        LOG.info("Failed to invoke the endpoint. Url: " + endpointURL + ". Retrying the request.");
+                        result = executeRequestWithRetries(request, endpointURL, requestRetryCount);
+                    }
+                    outcome = result.getRight().getLeft();
+                    JSONObject json = result.getRight().getRight();
+                    asyncReturn.accept(context, json != null ? json : Collections.emptyMap(), outcome);
+                }
+            } catch (Exception e) {
+                LOG.error("Error while applying authentication to the request.", e);
+                asyncReturn.accept(context, Collections.emptyMap(), Constants.OUTCOME_FAIL);
             }
+        });
+        JsGraphBuilder.addLongWaitProcess(asyncProcess, eventHandlers);
+    }
 
-            if (!isValidRequestDomain(request.getURI())) {
-                outcome = Constants.OUTCOME_FAIL;
-                LOG.error("Provided Url does not contain a allowed domain. Invalid Url: " + endpointURL);
-                asyncReturn.accept(context, Collections.emptyMap(), outcome);
-                return;
+    /**
+     * Execute the request with retries.
+     *
+     * @param request     HttpUriRequest.
+     * @param endpointURL Endpoint URL.
+     * @param maxRetries  Maximum number of retries.
+     * @return Pair of outcome and json.
+     */
+    private Pair<RetryDecision, Pair<String, JSONObject>> executeRequestWithRetries
+    (HttpUriRequest request, String endpointURL, int maxRetries) {
+
+        Pair<RetryDecision, Pair<String, JSONObject>> result;
+        String outcome = Constants.OUTCOME_FAIL;
+        int attempts = 0;
+        RetryDecision isRetry = RetryDecision.NO_RETRY;
+
+        while (attempts < maxRetries) {
+            attempts++;
+            LOG.warn("Retrying the request for endpoint: " + endpointURL + ". Attempt: " + attempts);
+            if (LoggerUtils.isDiagnosticLogsEnabled()) {
+                DiagnosticLog.DiagnosticLogBuilder diagnosticLogBuilder = new
+                        DiagnosticLog.DiagnosticLogBuilder(Constants.LogConstants.ADAPTIVE_AUTH_SERVICE,
+                        Constants.LogConstants.ActionIDs.RECEIVE_API_RESPONSE);
+                diagnosticLogBuilder.inputParam(Constants.LogConstants.InputKeys.API, endpointURL)
+                        .inputParam(Constants.LogConstants.InputKeys.ATTEMPT, attempts)
+                        .resultMessage("Retrying the request for external api.")
```

Comment:
Shall we add the retry attempt to the message rather than having it as an input field. 
`"Retrying the request for external api. Attempt: " + attemptCount + "."
`
It's not actually an input. I also noted you have added the attempt to the message when it comes to the retry request of the access token. So be consistent in both places.

Reply:
Resolved with https://github.com/wso2-extensions/identity-conditional-auth-functions/pull/157/commits/0da007ebe0bbc0242efba6c485f5d4f3a3cb2fe2